tensorflow/core/kernels/mkl_qmatmul_op.cc

```diff
@@ -0,0 +1,875 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// Implements a quantized eight-bit version of the matmul operation with bias,
+// relu and requantization fusion support utilizing mkldnn u8s8s32 inner
+// product API. Right now, this version can support
+//   - Input: quantized as uint8 via either MIN_FIRST or SCALE mode.
+//            SCALE mode is selected when input is guaranteed to be non-
+//            negative, e.g., MatMul is fed by Relu. Otherwise, MIN_FIRST is
+//            selected.
+//   - Weight: quantized to int8 via SCALE mode.
+//   - Bias: float32/int32. For int32, it is quantized according to input and
+//           filter min-max values.
+// Other than that, this op does not support other input combination yet.
+// When input is quantized to uint8 via MIN_FIRST, bias needs compensation.
+// The detailed algorithm is illustrated as below:
+//
+// A𝑓32 is the original fp32 activation 2D tensor.
+// Min(A𝑓32) is the minimum scalar value of A𝑓32.
+// Max(A𝑓32) is the maximum scalar value of A𝑓32.
+// MaxAbs(A𝑓32) is the maximum absolute scalar value of A𝑓32.
+// Qa is the quantization scale for activation.
+// Au8 is the quantized unsigned int8 activation tensor.
+// With SCALE quantization (used for non-negative A𝑓32), Qa and Au8 can be
+// calculated as below:
+//    Qa = 255.0 / Max(A𝑓32)
+//    Au8 = round(Qa * A𝑓32).
+// With MIN_FIRST quantization, Q'a and A'u8 can be calculated as below:
+//    Q'a = 255.0 / (Max(A𝑓32) – Min(A𝑓32))
+//    A'u8 = round(Q'a * (A𝑓32 – Min(A𝑓32) * ones(A𝑓32))),
+// where, ones(.) is a tensor of all 1s with the same shape of its argument and
+// round(.) rounds a number to its nearest integer.
+//
+// W𝑓32 is the original fp32 2D weight tensor.
+// MaxAbs(W𝑓32) is the maximum absolute scalar value of W𝑓32.
+// Qw is the quantization scale of weight.
+// Ws8 is the quantized signed int8 weight tensor.
+// Qw and Ws8 can be calculated as below:
+//    Qw = 127.0 / MaxAbs(W𝑓32)
+//    Ws8 = round(Qw * W𝑓32).
+//
+// B𝑓32 is the original fp32 1D bias tensor matching the innermost dim of W𝑓32.
+// With SCALE quantization of activation, the scaled bias, Bs32, is calculated
+// as below:
+//      Bs32 = Qa * Qw * B𝑓32.
+// With MIN_FIRST quantization of activation, the scaled bias tensor with
+// compensation, B's32, is calculated as below:
+//      B's32 = Q'a * Qw * B𝑓32 + Q'a * Qw * Min(A𝑓32) * 1 * W𝑓32
```

Comment:
Since you use `Ws8` in the actual implementation, please add an additional line
`= Q'a * Qw * B𝑓32 + Q'a * Min(A𝑓32) * 1 * Ws8`.